### DISCUSSION TITLE
Optimizations for numpy array communication?

### DISCUSSION THREAD

**Discussion by AleksBL** [mpi4py/mpi4py#195] [Q&A] [answered]
Hi 
I am working on some code that utilizes MPI4PY and having looked at the timings of the code i currently have, a significant part of the execution-time is spent communicating numpy arrays between processes. These arrays can be 100~10GB in size, so they are quite a chuck of data to communicate.

The question:

Are there any other ways of optimizing numpy-array communication, other using the capital methods of MPI4PY (Recv, Send)?

Best regards,
Aleksander

**Comment by dalcinl** [comment, marked as answer]
As long as you are using plain point-to-point (send/recv) operations, this may help a little: https://mpi4py.readthedocs.io/en/stable/mpi4py.util.pkl5.html

PS: Regarding collectives, as you can see, only `bcast()` is supported. I could implement scatter/[all]gather/alltoall using trivial p2p calls, but I haven't had any free time to do it. Another possibility for implementation would be to use user-defined struct datatypes with absolute addresses, and `MPI_Alltoallw()`, but again, I would need to experiment and I do not have any spare time right now.
